components/service/glean/src/main/java/mozilla/components/service/glean/storages/PingStorageEngine.kt

```diff
@@ -0,0 +1,178 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.storages
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import kotlinx.coroutines.launch
+import mozilla.components.service.glean.Dispatchers
+import mozilla.components.service.glean.Glean
+import java.io.BufferedReader
+import java.io.BufferedWriter
+import java.io.File
+import java.io.FilenameFilter
+import java.io.FileNotFoundException
+import java.io.FileOutputStream
+import java.io.FileReader
+import java.io.IOException
+import java.io.OutputStreamWriter
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+import java.util.UUID
+
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This class implementation stores pings as files on disk.
+ */
+internal open class PingStorageEngine(context: Context) {
+
+    private val logger: Logger = Logger("glean/PingStorageEngine")
+
+    @VisibleForTesting
+    internal val storageDirectory: File = File(context.applicationInfo.dataDir, PING_STORAGE_DIRECTORY)
+
+    companion object {
+        // Since ping file names are UUIDs, this matches UUIDs for filtering purposes
+        private const val FILE_PATTERN =
+            "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+        // Base directory for storing serialized pings
+        private const val PING_STORAGE_DIRECTORY = "${Glean.GLEAN_DATA_DIR}/ping_storage"
+    }
+
+    /**
+     * Function to store the ping to file for the [PingUploadWorker] to read and send when it is run
+     *
+     * @param pingType String representing the ping type: baseline, events, metrics (or custom ping)
+     * @param pingPath String representing the upload path used for uploading the ping
+     * @param ping Serialized JSON string representing the ping payload
+     */
+    @ObsoleteCoroutinesApi
+    @Synchronized
+    fun store(pingType: String, uuidFileName: UUID, pingPath: String, ping: String) {
+        Dispatchers.API.launch {
+            val pingStorageDirectory = File(storageDirectory, pingType)
+            assertDirectory(pingStorageDirectory)
+
+            // Build the file path for the ping, using the UUID from the path for the file name
+            val pingFile = File(pingStorageDirectory, uuidFileName.toString())
+
+            var stream: FileOutputStream? = null
+
+            try {
```

Comment:
Instead of using try, we could use kotlin-use. See [here](https://stackoverflow.com/a/26979481/261698)

Reply:
I think we still want the try/catch for the call to `write()`, but I have no problem with doing this with `.use {}`